### PR TITLE
Tank Source Side Flow Control Mode IndirectHeatAlternateSetpoint Code Not Hit

### PR DESCRIPTION
**Pull request by yzhou601**
Pull request overview
---------------------
The source side flow set by calling function ``PlantMassFlowRatesFunc`` didn't actually work for source side flow control mode: IndirectHeatAlternateSetpoint for indirect water heater system and tankless coils. Relevant question on Unmet Hour: [Click Here](https://unmethours.com/question/30394/indirectly-heated-water-heater-with-a-backup-electric-heater-indirectheatalternatesetpoint/)
The [issue](https://github.com/NREL/EnergyPlus/issues/3947) and [CR9176](https://github.com/NREL/EnergyPlusArchive/commit/dfceb055639261df4354f7b9967d82dee1c9dc71) changes are made under condition of ``MaybeRequestingFlow`` which is not stepped at final iteration in every time step. 
So these codes are added to condition ``ThrottlingFlow`` to fix this problem.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by yzhou601** [comment]
This bug fix issue seems not have been covered in tests without any CI test fail. So it's necessary to show my local test results for indirect water heater system models (no backup water heater, 55C alternate setpoint schedule, 52.667C WH Setpoint, 1C deadband).

I expected the source side mass flow rate to be over 0 for time points when tank temperature is below 55C-1C. 

### Run in Develop

#### IndirectHeatAlternateSetpoint

image (4)

#### IndirectHeatPrimarySetpoint

image (6)

### Run in this branch

#### IndirectHeatAlternateSetpoint

image (5)

#### IndirectHeatPrimarySetpoint

image (7)

It shows more insights of this bug, the alternate option is not really working.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterThermalTanks.cc]: I very much like this duplicated code is extracted out into a worker function.  The format is messy here though, need to include spaces after each comma in an argument list.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterThermalTanks.cc]: If you aren't going to actually include a methodology, just eliminate these placeholders/stubs.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterThermalTanks.cc]: I'd prefer you actually include `ScheduleManager::GetCurrentScheduleValue` in the code below rather than having a `using` statement at the top of the function.  In this case, it's not a big deal, but in many places within E+, the usage of `using` is huge and causes uncertainty around where a function or variable is being accessed.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterThermalTanks.cc]: These can be defined at the point of usage since they are only used within that scope, and also declared `const` for clarity.

File: tst/EnergyPlus/unit/WaterThermalTanks.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/WaterThermalTanks.unit.cc]: I was expecting this unit test to include a direct call to `SourceHeatNeed` since you extracted it out into a worker function, but I guess it's called from within other calls below.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterThermalTanks.cc]: A nice clean up here would be to either:
- Make this function a method on the tank data structure, that way you could just call `WaterThermalTank(WaterThermalTankNum).sourceHeatNeed(outletTemp, dbTemp, spTemp)` and eliminate the array lookups, or
- Pass in a const tank object as an argument instead of the tank num, again eliminating the array lookups.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
The functional changes here look good, but I'd like to see some improvements made before this gets merged.  Take a look at my comments and let me know what you think.

File: src/EnergyPlus/WaterThermalTanks.cc
Comment (yzhou601) [reply src/EnergyPlus/WaterThermalTanks.cc]: Thanks, Edwin, I passed the tank object as SourceHeatNeed function argument (option 2)

File: tst/EnergyPlus/unit/WaterThermalTanks.unit.cc
Comment (yzhou601) [reply tst/EnergyPlus/unit/WaterThermalTanks.unit.cc]: Yes it's calling ``PlantMassFlowRatesFun`` inside ``InitWaterThermalTank`` which is not straghtforward. It's a good idea to directly hit this function. I addressed this in the new commits below.

**Comment by yzhou601** [comment]
~~@Myoldmopar Do you know why these commits that happened in develop branch all appeared in this PR once I pulled develop into this branch?~~
(seemed to be normal again.)

**Comment by Myoldmopar** [comment]
Yeah I don't know what happened there.  But your [develop](https://github.com/yzhou601/EnergyPlus/tree/develop) branch is 8 commits ahead of the NREL/EnergyPlus develop branch.  So that's potentially problematic, but it seems not with the current issue.

**Comment by yzhou601** [comment]
> Yeah I don't know what happened there. But your [develop](https://github.com/yzhou601/EnergyPlus/tree/develop) branch is 8 commits ahead of the NREL/EnergyPlus develop branch. So that's potentially problematic, but it seems not with the current issue.

Yeah, these 8 commits are all pulling the latest changes from NREL/EnergyPlus/develop, might be fine?

**Comment by Myoldmopar** [comment]
No, your develop branch should never be _ahead_ of the NREL develop branch.  Pulling in the latest changes from NREL develop should cause a git fast-forward, never a merge, so it should never be ahead.  Something went wrong.  But again, I don't think it caused an issue here, just something you need to get addressed before it accidentally does.

**Comment by mjwitte** [comment]
@Myoldmopar Um, there hasn't been a commit to this branch for 6 or 7 days, yet CI keeps making the rounds on this one. Working on it again now.

**Comment by Myoldmopar** [comment]
Yeah, CI has been having a bit of trouble with some external PRs lately.  Nothing serious, just wastes some cycles running extra commits.  I'm currently unclear as to the reason, but will dig into it soon after the release, if not before.

File: tst/EnergyPlus/unit/WaterThermalTanks.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/WaterThermalTanks.unit.cc:1995]: I'm not sure what editor you are using to write C++, but it shouldn't be leaving these trailing whitespaces in the files.  In most editors there is an option to clean any trailing whitespace when you save a file, I'd look for that option.  I'm not going to hold up this PR just for that though.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Code changes look much better now.  I'm testing locally.

**Comment by Myoldmopar** [comment]
Merged up with develop, unit test suite passes, and ran a sampling of regression files and got no diffs.  I'm merging this in.

**Pull request merged**
